atintegrators/VariableThinMPolePass.c

```diff
@@ -0,0 +1,274 @@
+/* VariableThinMPolePass
+   Accelerator Toolbox 
+   S.White simon.white@esrf.fr
+*/
+
+#include "atelem.c"
+#include "atlalib.c"
+#include "driftkick.c"
+
+#define TWOPI  6.28318530717959
+
+
+struct elemab
+{
+    double *Amplitude;
+    double Frequency;
+    double Phase;
+    int NSamples;
+    double *Func;
+};
+
+struct elem 
+{
+    double *PolynomA;
+    double *PolynomB;
+    struct elemab *ElemA;
+    struct elemab *ElemB;
+    int Seed;
+    int Mode;
+    int MaxOrder;
+    double *Ramps;
+};
+
+
+double randn(int seed)
+{
+  double u, v, w, mult;
+  static int initseed=1;
+
+  if(initseed){
+      srand(seed);
+      initseed=0;
+  }
+
+  do{
+      u = -1 + ((double) rand () / RAND_MAX) * 2;
+      v = -1 + ((double) rand () / RAND_MAX) * 2;
+      w = u*u + v*v;
+  }
+  while (w >= 1 || w == 0);
+  mult=sqrt ((-2 * log (w)) / w);
+  return u*mult;
+}
+
+double get_amp(double amp, double *ramps, double t)
+{
+    double ampt=amp;
+    if(ramps){
+        if(t<=ramps[0]){
+            ampt=0.0;
+        }else if(t<=ramps[1]){
+            ampt=amp*(t-ramps[0])/(ramps[1]-ramps[0]);
+        }else if(t<=ramps[2]){
+            ampt=amp;
+        }else if(t<=ramps[3]){
+            ampt=amp-amp*(t-ramps[2])/(ramps[3]-ramps[2]);
+        }else{
+            ampt=0.0;
+        }
+    }
+    return ampt;
+}
+
+double get_pol(struct elemab *elem, double *ramps, int mode,
+               double t, int turn, int seed, int order)
+{
+    int idx;
+    double ampt, freq, ph, val;
+    double *func;
+    double *amp = elem->Amplitude;
+    if(!amp){
+        return 0.0;
+    }
+    ampt=get_amp(amp[order],ramps,turn);
+    switch(mode){
+    case 0:
+        freq = elem->Frequency;
+        ph = elem->Phase;
+        ampt *= sin(TWOPI*freq*t+ph);
+        return ampt;
+    case 1:
+        val = randn(seed);
+        ampt *= val;
+        return ampt;
+    case 2:
+        func = elem->Func;
+        idx = turn%elem->NSamples;
+        ampt *= func[idx];
+        return ampt;
+    }

```

Comment:
The compiler complains that the function `get_pol` may return nothing (`void`) instead of `double` if `mode` is not 0, 1, or 2. This should never happen, but one could avoid the warning either by adding a `default` clause or by returning 0.0 after the end of the switch

Reply:
Done